### PR TITLE
New mail motor module

### LINKED ISSUE #1164: The New Mailmotor
Hi,

At the [Last Fork CMS Meetup](http://www.meetup.com/Fork-CMS/events/220501469/) I talked to some Fork CMS developers (@diewy, @thijzer, @ikoene) about refactoring the Mailmotor to support any Mail engine available.

Since I already tried to [refactor mailmotor to support multiple mail engines like f.e. Mailchimp, next to the already integrated CampaignMonitor](https://github.com/siesqo/fork-cms-module-mailmotor). I was almost finished integrating everything, but I still came across to [unsolvable problems/issues](https://github.com/siesqo/fork-cms-module-mailmotor/issues) like f.e. non-existing functionality and the biggest part - maintainability.

So this is my call to all @forkcms/owners and @forkcms/moderators if we could talk about reducing the mailmotor to it simple basic functionality like:
- Subscribe/Unsubscribe subscribers from any external mail engine

And not:
- Storing subscribers in our own database
- Creating and sending mails. Which is something that can be discussed about.
  Yes, it is so cool if the client can do this, but the downside is maintainability and the very low use case.
  It is often used by marketing to pull new clients in. But it is rarely used.

Why downsizing?

We would attract more customers - willing to try Fork CMS - since they can use the mail engine they like.

What do you think about this?
Your opinion will be fully appreciated.
# 

Features overview:

**Frontend**
- [x] Subscribe widget
- [x] Unsubscribe widget

**Backend**
- Dashboard
  - [ ] Widget for subscriptions/unsubscriptions
  - [ ] v2: Widget for list growth
- Mailmotor
  - [ ] v2: Statistics for campaigns.
- Settings
  - [ ] Checkbox: "Add a first/last name field in Subscribe form"

### PR DESCRIPTION
**Pull request by jeroendesloovere**
Fixes #1164

### CODE REVIEW

**Comment by carakas** [comment]
CI fails

File: app/config/parameters.yml.dist [hunk 1]
```diff
@@ -23,6 +23,6 @@ parameters:
     action.group_tag:       <action-group-tag>
     action.rights_level:    <action-rights-level>
 
-    mailmotor.mail_engine:  'mailchimp'
+    mailmotor.mail_engine:  'not_implemented'
```
Comment (carakas) [inline app/config/parameters.yml.dist]: This is confusing

File: src/Backend/Modules/MailMotor/Actions/Settings.php [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+     *
+     * @return array(key => value)
+     */
+    private function getDropdownValuesForMailEngines()
+    {
+        // init dropdown values
+        $ddmValuesForMailEngines = array();
+
+        // loop all container services to find "mail-engine" gateway services
+        foreach ($this->getContainer()->getServiceIds() as $serviceId) {
```
Comment (carakas) [inline src/Backend/Modules/MailMotor/Actions/Settings.php]: you should use service tags for this

File: src/Backend/Modules/MailMotor/Actions/Settings.php [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+        return $ddmValuesForMailEngines;
+    }
+
+    /**
+     * Load form
+     */
+    public function loadForm()
+    {
+        $this->form = new Form('settings');
```
Comment (carakas) [inline src/Backend/Modules/MailMotor/Actions/Settings.php]: can you change this to symfony form please

File: src/Backend/Modules/MailMotor/Config.php [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+ * @author Jeroen Desloovere <jeroen@siesqo.be>
+ */
+class Config extends BackendBaseConfig
+{
+    /**
+     * The default action
+     *
+     * @var	string
+     */
+    protected $defaultAction = '';
```
Comment (carakas) [inline src/Backend/Modules/MailMotor/Config.php]: this shouldn't be empty

File: src/Backend/Modules/MailMotor/DependencyInjection/Compiler/CustomCompilerPass.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
+<?php
+
+namespace Backend\Modules\MailMotor\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+
+class CustomCompilerPass implements CompilerPassInterface
```
Comment (carakas) [inline src/Backend/Modules/MailMotor/DependencyInjection/Compiler/CustomCompilerPass.php]: can you give this a better name than CustomCompilerPass

File: src/Backend/Modules/MailMotor/DependencyInjection/MailMotorExtension.php [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+ * This is the class that loads and manages your bundle configuration
+ *
+ * To learn more see {@link http://symfony.com/doc/current/cookbook/bundles/extension.html}
+ */
+class MailMotorExtension extends Extension
+{
+    /**
+     * {@inheritDoc}
+     */
+    public function load(array $configs, ContainerBuilder $container)
```
Comment (carakas) [inline src/Backend/Modules/MailMotor/DependencyInjection/MailMotorExtension.php]: why did you add this class if it is empty

File: src/Backend/Modules/MailMotor/Installer/Installer.php [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        }
+    }
+
+    /**
+     * Install settings
+     */
+    private function installSettings()
+    {
+        // add 'MailMotor' as a module
+        $this->addModule('MailMotor', 'The module that allows you to insert/delete subscribers to/from your mailinglist.');
```
Comment (carakas) [inline src/Backend/Modules/MailMotor/Installer/Installer.php]: after adding the module you can use `$this->getModule()` instead of MailMotor

File: src/Backend/Modules/MailMotor/Actions/Settings.php [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+use Backend\Core\Engine\Form;
+use Backend\Core\Engine\Language;
+use Backend\Core\Engine\Model;
+use Symfony\Component\Filesystem\Filesystem;
+
+/**
+ * This is the settings-action (default), 
+ * it will be used to couple your "mail-engine" account
+ *
+ * @author Jeroen Desloovere <jeroen@siesqo.be>
```
Comment (carakas) [inline src/Backend/Modules/MailMotor/Actions/Settings.php]: we no longer have author tags in fork in favour of `git blame`

File: src/Backend/Modules/MailMotor/Js/Mailmotor.js [hunk 9]
```diff
@@ -0,0 +1,36 @@
+/**
+ * Interaction for the MailMotor module
+ *
+ * @author Jeroen Desloovere <jeroen@siesqo.be>
+ */
+jsBackend.mailmotor =
+{
+	// init, something like a constructor
```
Comment (carakas) [inline src/Backend/Modules/MailMotor/Js/Mailmotor.js]: silly comment

File: src/Backend/Modules/MailMotor/Js/Mailmotor.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+	}
+};
+
+jsBackend.mailmotor.controls =
+{
+	// init, something like a constructor
+	init: function()
+	{
+		// stop here because mail engine not found
+		if ($('#mailEngine').length == 0) {
```
Comment (carakas) [inline src/Backend/Modules/MailMotor/Js/Mailmotor.js]: should be `===`

File: src/Backend/Modules/MailMotor/Js/Mailmotor.js [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+			return false;
```
Comment (carakas) [inline src/Backend/Modules/MailMotor/Js/Mailmotor.js]: you don't use the boolean so you can just return

File: src/Backend/Modules/MailMotor/Layout/Templates/Settings.tpl [hunk 12]
```diff
@@ -0,0 +1,60 @@
+{include:{$BACKEND_CORE_PATH}/Layout/Templates/Head.tpl}
+{include:{$BACKEND_CORE_PATH}/Layout/Templates/StructureStartModule.tpl}
+
```
Comment (carakas) [inline src/Backend/Modules/MailMotor/Layout/Templates/Settings.tpl]: you should update this template to twig

File: src/Frontend/Modules/MailMotor/Actions/Subscribe.php [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+ * This is the Subscribe-action for our MailMotor
+ *
+ * @author Jeroen Desloovere <jeroen@siesqo.be>
+ */
+class Subscribe extends FrontendBaseBlock
+{
+    /**
+     * FrontendForm instance
+     *
+     * @var	FrontendForm
```
Comment (carakas) [inline src/Frontend/Modules/MailMotor/Actions/Subscribe.php]: don't mix tabs and spaces

File: src/Frontend/Modules/MailMotor/Actions/Subscribe.php [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+
+    /**
+     * Load the form
+     *
+     * @return void
+     */
+    private function loadData()
+    {
+        try {
+            // Define interests
```
Comment (carakas) [inline src/Frontend/Modules/MailMotor/Actions/Subscribe.php]: useless comment

File: src/Frontend/Modules/MailMotor/Actions/Subscribe.php [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+        parent::execute();
+        $this->loadData();
+        $this->loadTemplate();
+        $this->loadForm();
+        $this->validateForm();
+        $this->parse();
+    }
+
+    /**
+     * Load the form
```
Comment (carakas) [inline src/Frontend/Modules/MailMotor/Actions/Subscribe.php]: wrong comment

File: src/Frontend/Modules/MailMotor/Actions/Subscribe.php [hunk 16]
```diff
@@ -0,0 +1,252 @@
...
+     * @return void
+     */
+    private function loadData()
+    {
+        try {
+            // Define interests
+            $this->interests = $this->get('mailmotor.subscriber')->getInterests();
+        // Fallback for when no mail-engine is chosen in the Backend
+        } catch (NotImplementedException $e) {
+            $this->interests = false;
```
Comment (carakas) [inline src/Frontend/Modules/MailMotor/Actions/Subscribe.php]: you are mixing an array with a bool
I would set this to an empty array

File: src/Frontend/Modules/MailMotor/Actions/Subscribe.php [hunk 17]
```diff
@@ -0,0 +1,252 @@
...
+                array(
+                    'required' => null,
+                    'type' => 'email',
+                    'placeholder' => ucfirst(Language::lbl('YourEmail')),
+                )
+            )
+        ;
+
+        // Has interests
+        if ($this->interests) {
```
Comment (carakas) [inline src/Frontend/Modules/MailMotor/Actions/Subscribe.php]: you should check here if the intrests array isn't empty instead of mixing types with a bool

File: src/Frontend/Modules/MailMotor/Actions/Subscribe.php [hunk 18]
```diff
@@ -0,0 +1,252 @@
...
+        // request contains an email
+        if ($this->get('request')->request->get('email') != null) {
+            $email = $this->get('request')->request->get('email');
+        }
+
+        // create & add elements
+        $this->frm->addText('email', $email);
+        $this->frm->addText('email')
+            ->setAttributes(
+                array(
```
Comment (carakas) [inline src/Frontend/Modules/MailMotor/Actions/Subscribe.php]: you use the new array syntax on other places, you shouldn't mix it

File: src/Frontend/Modules/MailMotor/Actions/Subscribe.php [hunk 19]
```diff
@@ -0,0 +1,252 @@
...
+
+    /**
+     * Parse the data into the template
+     *
+     * @return void
+     */
+    protected function parse()
+    {
+        // form was sent?
+        if ($this->URL->getParameter('sent') == 'true') {
```
Comment (carakas) [inline src/Frontend/Modules/MailMotor/Actions/Subscribe.php]: `===`

Comment (carakas) [inline src/Frontend/Modules/MailMotor/Actions/Subscribe.php]: using sent is a bad practice since it is too generic and may clash with other widgets on the page

File: src/Frontend/Modules/MailMotor/Engine/Model.php [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+
+use Frontend\Core\Engine\Language;
+use Frontend\Core\Engine\Model as FrontendModel;
+
+/**
+ * In this file we store all generic functions that we will be using in the MailMotor module
+ *
+ * @author Jeroen Desloovere <jeroen@siesqo.be>
+ */
+class Model
```
Comment (carakas) [inline src/Frontend/Modules/MailMotor/Engine/Model.php]: I would remove the model end do all this with events

File: src/Frontend/Modules/MailMotor/Layout/Templates/Subscribe.tpl [hunk 21]
```diff
@@ -0,0 +1,48 @@
+{*
+	The variables that are available to use:
```
Comment (carakas) [inline src/Frontend/Modules/MailMotor/Layout/Templates/Subscribe.tpl]: spoon templates should be removed

File: src/Backend/Modules/MailMotor/Actions/Settings.php (hunk 3 above)
Comment (jeroendesloovere) [inline src/Backend/Modules/MailMotor/Actions/Settings.php]: God dammed :-P
